### ISSUE TITLE
False positive: error CS8352: Cannot use local 'xxx' in this context because it may expose referenced variables outside of their declaration scope

### ISSUE THREAD

**Issue by bernd5** [dotnet/roslyn#53014]
**Version Used**: [Link](https://github.com/dotnet/roslyn/commit/4e4c8237665adb00a0d0c5f710b9f5ce1a2086d6)

**Steps to Reproduce**:
[Compile:](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAgrgOwD4AEBMAjALABQZAygA4SYA8AlpjAHwAEm6YwiUbAvG1gQAxgGsIAG0kB7EWyYwA2gDYAugG4ynbryUAGNQLaEtpHTwRQlhI4PxmLe/HbYBmR10vW3rgCyeulZKfq4ArIHeSmGuKmbUtIzM7FAQYHAJmGap6ZnGTlYaQA===)

```cs
using System;

Span<int> numbers = stackalloc int[6];
numbers[0] = 1;
numbers[1] = 2;
numbers[2] = 3;
numbers[3] = 4;
numbers[4] = 5;
numbers[5] = 6;
Span<int> sameSpan;
sameSpan = numbers;
```

**Expected Behavior**:
No error

**Actual Behavior**:
error CS8352: Cannot use local 'numbers' in this context because it may expose referenced variables outside of their declaration scope

-----------------
If sameSpan would be declared in a parent scope the error would be correct.

**Comment by alrz** [comment]
This is currently by-design, see https://github.com/dotnet/csharplang/discussions/1130

> ```cs
> Span<T> s;
> ```
> is the same as
> ```cs
> Span<T> s = default;
> ```

I think we should not deduce Span lifetime at the time of declaration, rather it should happen at the point of assignment, that way both these issues could be resolved.

**Comment by bernd5** [comment]
Ah, it is because of a life-time "detection issue". `Span<T>`'s lifetime ends at the end of the current scope, regardless of any assignment.
If a ref struct would be assigned to a variable which has a larger scope - such an error must be emitted - otherwise not.
And of course it can't be returned (lifetime ends before return).

As a workaround we can just ignore those checks and write something like:
```cs
Span<int> numbers = stackalloc int[6];
numbers[0] = 1;
numbers[1] = 2;
numbers[2] = 3;
numbers[3] = 4;
numbers[4] = 5;
numbers[5] = 6;
Span<int> sameSpan;
unsafe
{
    sameSpan = new Span<int>(global::System.Runtime.CompilerServices.Unsafe.AsPointer(ref numbers[0]), numbers.Length);
}
Console.WriteLine(sameSpan[0]);
```

But that need a `unsafe` block.

**Comment by bernd5** [comment]
I assume this has the same reason why the [following ](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAgrgOwD4FgBQAlpjAASGkC8pAjAEwDcBCcAZuSeS+9d+U/j7EyUCGDgVebUgHMANgHtgEOSBAABGgDoASlhiFxWgMIKwAB0JzEAZUQA3QgGM4ULQFVMo1nC0A5dDk5HTYAHmEAPgAKAEoBUXFJUiEBIA=)does not work, too:

```cs
int i = 12;
ref int iref = ref i;
ref int samei = ref global::System.Runtime.CompilerServices.Unsafe.NullRef<int>();
samei = ref i;
```

=> error CS8374: Cannot ref-assign 'i' to 'samei' because 'i' has a narrower escape scope than 'samei'.

**Comment by HobbsCode** [comment]
If you need to decouple the declaration and assignment of the second span, this will work. 
```
Span<int> numbers = stackalloc int[6];
numbers[0] = 1;
numbers[1] = 2;
numbers[2] = 3;
numbers[3] = 4;
numbers[4] = 5;
numbers[5] = 6;
        
Span<int> sameSpan = stackalloc int[0];
sameSpan = numbers;
```

**Comment by bernd5** [comment]
@HobbsCode
thanks - this is a good workaround.
